src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/RemoteDebuggerAuthenticationService.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Composition;
+using Microsoft.VisualStudio.ProjectSystem.Debug;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Debug
+{
+    [Export(typeof(IRemoteDebuggerAuthenticationService))]
+    [AppliesTo(ProjectCapabilities.AlwaysApplicable)]
+    internal class RemoteDebuggerAuthenticationService : IRemoteDebuggerAuthenticationService
+    {
+        private readonly IVsUIService<SVsDebugRemoteDiscoveryUI, IVsDebugRemoteDiscoveryUI> _remoteDiscoveryUIService;
+        private readonly IProjectThreadingService _threadingService;
+
+        [ImportingConstructor]
+        public RemoteDebuggerAuthenticationService(
+            UnconfiguredProject unconfiguredProject,
+            IVsUIService<SVsDebugRemoteDiscoveryUI, IVsDebugRemoteDiscoveryUI> remoteDiscoveryUIService,
+            IProjectThreadingService threadingService)
+        {
+            _remoteDiscoveryUIService = remoteDiscoveryUIService;
+            _threadingService = threadingService;
+            AuthenticationProviders = new OrderPrecedenceImportCollection<IRemoteAuthenticationProvider>(projectCapabilityCheckProvider: unconfiguredProject);
+        }
+
+        [ImportMany]
+        private OrderPrecedenceImportCollection<IRemoteAuthenticationProvider> AuthenticationProviders { get; }
+
+        public IRemoteAuthenticationProvider? GetProviderForAuthenticationMode(string? remoteAuthenticationMode) => AuthenticationProviders.FirstOrDefaultValue(p => p.Name.Equals(remoteAuthenticationMode, StringComparisons.LaunchProfileProperties));
```

Comment:
Why is null/empty allowed?

Reply:
Assuming you mean for the parameter, empty could be in the json file, or come from a bad json file, and null could come from the ConsoleDebugTargets if the profile isn't full defined - but also because string.Equals handles all of those cases happily.

Reply:
Nit: APIs that return "null" to help the caller do a null check make it harder to understand the API itself. It gives the impression that `remoteAuthenticationMode` is optional, when in fact its not if you want anything useful out of it.

Reply:
I'm not sure I follow this one. If you're talking about the return value, that allows nulls because this is processing user input from the json file so it could be rubbish data. Would a TryGetXXX form make more sense to you? If you're talking about the parameter, again thats null because it could be null in the JSON file. This could disallow nulls, but that would mean the thing that reads the JSON file needs to know that a null should be mapped to an empty string, or avoid calling this method, which might not be something it can reason about. Maybe its just me but I don't equate nullable with optional, since we have optional parameters. Nullable just means that the method can deal with a null ¯\_(ツ)_/¯

Reply:
Providers cannot have a null name, yet this method takes a nullable name. That is my point.

It does this so that the JSON parsing code doesn't need to do a null check before calling this method.

Reply:
Why would the JSON code need to map null -> empty? Empty names also do not make sense.